src/server.ts

```diff
@@ -0,0 +1,232 @@
+  const { props, type } = child;
+  let ownProps = assign(defaultProps, props);
+  if (type && type.defaultProps) ownProps = assign(defaultProps, type.defaultProps, props);
+  return ownProps;
+}
+
+export function getChildFromComponent(component) {
+  // See if this is a class, or stateless function
+  if (component && component.render) return component.render();
+  return component;
+}
+
+export function processQueries(queries, client): Promise<any> {
+  queries = flatten(queries)
+    .map((queryDetails: any) => {
+      const { query, component, ownProps, key } = queryDetails;
+      return client.query(query)
+        .then(result => {
+          const { data, errors } = result as any;
+          ownProps[key] = assign({ loading: false, errors }, data);
+          return {
+            component,
+            ownProps: assign({}, ownProps),
+          };
+        });
+    });
+
+  return Promise.all(queries);
+}
+
+function getQueriesFromTree({
+  components,
+  queries = [],
+  defaultProps = {},
+  context = {},
+}: QueryTreeArgument) {
+  Children.forEach(components, (child: any) => {
+    if (!child) return;
+    // XXX make the context dynamic
+    let { store, client } = context as Context;
+
+    // get the store
+    // XXX determine if this is actually a store
+    if (!store && child.props && child.props.store) {
+      context.store = store;
+      delete child.props.store;
+    }
+
+    let Element;
+    // find the client in the tree
+    if (!client && child.props && child.props.client instanceof ApolloClient) {
+      context.client = child.props.client as ApolloClient;
+      // delete child.props.client;
+      // for apps that don't provide their own store
+      if (!store) {
+        context.client.initStore();
+        context.store = context.client.store;
+      }
+    }
+
+    let ownProps = getPropsFromChild(child, defaultProps);
+    let state = store ? store.getState() : {};
+
+    // see if this is a connect type
+    if (child.type && typeof child.type.mapQueriesToProps === 'function') {
+      const dataRequirements = child.type.mapQueriesToProps({ ownProps, state });
+      for (let query in dataRequirements) {
+        ownProps[query] = assign({}, defaultReactProps);
+        if (dataRequirements[query].ssr === false) continue; // don't run this on the server
+
+        queries.push({
+          key: query,
+          query: dataRequirements[query],
+          component: child.type.WrappedComponent,
+          ownProps,
+        });
+      }
+
+      Element = createElement(child.type.WrappedComponent, ownProps) as any;
+    }
+
+    // try to see if this is a component, or a stateless component
+    if (!Element && typeof child.type === 'function') Element = { type: child.type };
+    const RenderedComponent = Element && Element.type && new Element.type(ownProps, context);
+
+    if (RenderedComponent && RenderedComponent.context) context = RenderedComponent.context;
```

Comment:
Are we overriding the `context` that's returned from the function? I'm confused about `context` here. Is it the same thing as `getChildContext` and co use?